File: plugins/cordova-plugin-muzik/src/android/GestureBridge.java
package com.muzik.androidplugin;

import com.muzik.accessory.MzAccessory;
import com.muzik.accessory.callback.IMzGestureCallback;
import com.muzik.accessory.MzGesture;

import java.util.ArrayList;
import java.util.HashMap;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;

public class GestureBridge implements IMzGestureCallback {

    private MzAccessory mzAccessory;
    private HashMap<Integer, CallbackContext> callbacks;

    public GestureBridge(MzAccessory mzAccessory) {
        this.mzAccessory = mzAccessory;
        this.callbacks = new HashMap<Integer, CallbackContext>();
    }

    public synchronized void registerForGestures(CallbackContext callback, int... gestures) {
        for(int gesture : gestures) {
            this.callbacks.put(gesture, callback);
        }
        MzGesture[] gesturesToRegister = new MzGesture[gestures.length];
        for (int i = 0; i < gesturesToRegister.length; i++) {
            gesturesToRegister[i] = getMuzikGestureForIndex(gestures[i]);
        }
        this.mzAccessory.registerForGestures(this, gesturesToRegister);
    }

    public synchronized void registerForAllGestures(CallbackContext callback, boolean includePassthroughGestures) {
        ArrayList<Integer> gesturesToRegister = new ArrayList<Integer>();
        for (MzGesture gesture : MzGesture.values()) {
            if (!gesture.isPassthrough() || includePassthroughGestures) {
                gesturesToRegister.add(gesture.getValue());
            }
        }

        int[] gestures = new int[gesturesToRegister.size()];
        for (int i = 0; i < gestures.length; i++) {
            gestures[i] = gesturesToRegister.get(i);
        }
        this.registerForGestures(callback, gestures);
    }

    public synchronized void unregisterForGestures(int... gestures) {
        for(int gesture : gestures) {
            this.callbacks.remove(gesture);
        }
        MzGesture[] gesturesToUnregister = new MzGesture[gestures.length];
        for (int i = 0; i < gesturesToUnregister.length; i++) {
            gesturesToUnregister[i] = getMuzikGestureForIndex(gestures[i]);
        }
        this.mzAccessory.unregisterForGestures(gesturesToUnregister);
    }

    public synchronized void unregisterForAllGestures(boolean includePassthroughGestures) {
        ArrayList<Integer> gesturesToUnregister = new ArrayList<Integer>();
        for (MzGesture gesture : MzGesture.values()) {
            if (!gesture.isPassthrough() || includePassthroughGestures) {
                gesturesToUnregister.add(gesture.getValue());
            }
        }

        int[] gestures = new int[gesturesToUnregister.size()];
        for (int i = 0; i < gestures.length; i++) {
            gestures[i] = gesturesToUnregister.get(i);
        }
        this.unregisterForGestures(gestures);
    }

    @Override
    public synchronized void onResponseReceived(MzGesture gesture, byte data[]) {
        CallbackContext callback = callbacks.get(gesture.getValue());
        if (callback != null) {
            PluginResult result = new PluginResult(PluginResult.Status.OK, gesture.getValue());
            result.setKeepCallback(true);
            callback.sendPluginResult(result);
        }
    }

    private MzGesture getMuzikGestureForIndex(int gestureIndex)
    {
        for (MzGesture gesture : MzGesture.values())
        {
            if (gesture.getValue() == gestureIndex)
            {
                return gesture;
            }
        }

        return null;
    }
}

File: cordova-plugin-muzik/src/ios/src/GestureBridge.m
//
//  GestureBridge.h
//

#import "GestureBridge.h"

@interface GestureBridge ()

@property (nonatomic, strong) MZAccessory* mzAccessory;
@property (nonatomic, strong) CDVPlugin* plugin;
@property (nonatomic, strong) NSMutableDictionary* callbacks;

@end

@implementation GestureBridge

// Default constructor
- (id) init {
    self = [super init];
    if (self) {
        self.mzAccessory = nil;
        self.plugin = nil;
        self.callbacks = [[NSMutableDictionary alloc] init];
    }
    return self;
}

// Constructor with references to an instance of MZAccessory, and CDVPlugin
- (id) initWithAccessory:(MZAccessory *)mzAccessory plugin:(CDVPlugin*)plugin{
    self = [super init];
    if (self) {
        self.mzAccessory = mzAccessory;
        self.plugin = plugin;
        self.callbacks = [[NSMutableDictionary alloc] init];
    }
    return self;
}

// Register for gestures
- (void) registerForGestures:(CDVInvokedUrlCommand*)command {
    if (command == nil){
        return;
    }
    NSMutableArray<MZGesture*>* mzGestures = [[NSMutableArray<MZGesture*> alloc] init];
    NSArray* gestures = command.arguments;
    for (NSUInteger i = 0; i < gestures.count; ++i){
        NSUInteger index = [[gestures objectAtIndex:i] unsignedIntegerValue];
        MZGesture* gesture = [self getMuzikGestureForIndex:index];
        NSNumber* key = [NSNumber numberWithUnsignedInteger:index];
        [mzGestures addObject:gesture];
        [self.callbacks setObject:command forKey:key];
    }
    MZGestureCallback callback = ^void(MZGesture* gesture, Byte* data){
        CDVInvokedUrlCommand* gestureCommand = [self.callbacks objectForKey:gesture.value];
        CDVPluginResult* pluginResult = nil;
        if (gesture != nil){
            NSUInteger gestureIndex = [gesture.value unsignedIntegerValue];
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsNSUInteger:gestureIndex];
            [pluginResult setKeepCallbackAsBool:YES];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Invalid gesture"];
        }
        [self.plugin.commandDelegate sendPluginResult:pluginResult callbackId:gestureCommand.callbackId];
    };
    [self.mzAccessory registerForGestures:[mzGestures copy] withBlock:callback];
}

// Register for all gestures
- (void) registerForAllGestures:(CDVInvokedUrlCommand*)command {
    if (command == nil){
        return;
    }
    BOOL includePassthroughGestures = [[command.arguments objectAtIndex:0] boolValue];
    MZGestureCallback callback = ^void(MZGesture* gesture, Byte* data){
        CDVPluginResult* pluginResult = nil;
        if (gesture != nil){
            NSUInteger gestureIndex = [gesture.value unsignedIntegerValue];
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsNSUInteger:gestureIndex];
            [pluginResult setKeepCallbackAsBool:YES];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Invalid gesture"];
        }
        [self.plugin.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    };
    [self.mzAccessory registerForAllGesturesWithPassthroughs:includePassthroughGestures withBlock:callback];
}

// Unregister for gestures
- (void) unregisterForGestures:(CDVInvokedUrlCommand*)command {
    if (command == nil){
        return;
    }
    NSMutableArray<MZGesture*>* mzGestures = [[NSMutableArray<MZGesture*> alloc] init];
    NSArray* gestures = command.arguments;
    for (NSUInteger i = 0; i < gestures.count; ++i){
        NSUInteger index = [[gestures objectAtIndex:i] unsignedIntegerValue];
        MZGesture* gesture = [self getMuzikGestureForIndex:index];
        [mzGestures addObject:gesture];
    }
    [self.mzAccessory unregisterForGestures:[mzGestures copy]];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"unregisterForGestures success."];
    [self.plugin.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

// Unregister for all gestures
- (void) unregisterForAllGestures:(CDVInvokedUrlCommand*)command {
    if (command == nil){
        return;
    }
    BOOL passthroughs = [[command.arguments objectAtIndex:0] boolValue];
    NSString* message = [[NSString alloc] initWithFormat:@"unregisterForAllGestures success withPassthroughs: %@", (passthroughs ? @"YES" : @"NO")];
    [self.mzAccessory unregisterForAllGesturesWithPassthroughs:passthroughs];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:message];
    [self.plugin.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (MZGesture*) getMuzikGestureForIndex:(NSUInteger)index{
    for (MZGesture* gesture in MZGesture.values){
        NSUInteger gestureValue = [gesture.value unsignedIntegerValue];
        if (index == gestureValue){
            return gesture;
        }
    }
    return nil;
}

@end